### PR TITLE
Dependency Injection and Tests Using Fakes

### PR DESCRIPTION
**Pull request by bcarterterp**
Adds dependency injection into the project. [Here](https://themobilecoder.com/dependency-injection-in-flutter-using-riverpod/) is the documentation I referenced while building out Riverpod's dependency injection. I have also leveraged Riverpod's [code generation](https://pub.dev/packages/riverpod_generator) in order to streamline creating providers. Tests have been written to utilize Fakes over mocks. Itt allows ease of use, flexibility, and repeated use in unrelated tests. I do not think going forward we should remove Mockito for this app, sense it does provide others with the option to choose. Looking for any and all feedback!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,9 +17,19 @@ samples, guidance on mobile development, and a full API reference.
...
+
+### Notes
+- Since we are using Riverpod for our state management, we have decided to enable dependency injection through it. While most classes will use constructor injection as seeing in the [providers.dart] class, the Notifiers are different. Notifiers can grab any dependency throug their `ref` reference, which is kept and maintained by the `ProviderScope`.
+- Riverpod allows the ability to swap out dependencies for tests. This allows us the ability to use Fakes in our tests, and as close to implementation as we want. More on that in the Testing section.
+
+### Testing
+
+
```
Comment (iamricardo) [inline README.md:29]: Thoughts on removing this? or making it `# Testing` and having it move down to line 50?

File: README.md [hunk 2]
```diff
@@ -17,9 +17,19 @@ samples, guidance on mobile development, and a full API reference.
...
 ## Libraries and Frameworks Used
 
-- State Management: [Flutter Riverpod](https://riverpod.dev/)
+- State Management & Dependency Injection: [Flutter Riverpod](https://riverpod.dev/)
```
Comment (iamricardo) [inline README.md:32]: Could we add an entry for equatable?

**Review by iamricardo — COMMENTED** [review]
What are the thoughts on having folders within entity? `entity/event`, `entity/login`, `entity/request`, `entity/user`. Just wondering if it will get cluttered in the future. 

Mainly commenting on the things I understood at this point. Might have to skip some sections to get to riverpod to understand this better

File: README.md (hunk 2 above)
Comment (bcarterterp) [reply README.md:32]: Absolutely!

**Comment by bcarterterp** [comment]
> What are the thoughts on having folders within entity? `entity/event`, `entity/login`, `entity/request`, `entity/user`. Just wondering if it will get cluttered in the future.
> 
> Mainly commenting on the things I understood at this point. Might have to skip some sections to get to riverpod to understand this better

On directories for Entity. My thought is that an Entity will be used across screens. For instance User Information would be needed throughout the application. If an object is screen specific and influence's it's design, then I would put it with the state. Entities I see as a more global typing.

The Riverpod section is a touch outdated (We should make a note of this in the ramp-up), I would lean on their documentation as it is the most up to date.

Keep the Questions/Comments coming!

**Comment by cxw320** [comment]
> Adds dependency injection into the project. [Here](https://themobilecoder.com/dependency-injection-in-flutter-using-riverpod/) is the documentation I referenced while building out Riverpod's dependency injection. I have also leveraged Riverpod's [code generation](https://pub.dev/packages/riverpod_generator) in order to streamline creating providers. Tests have been written to utilize Fakes over mocks. Itt allows ease of use, flexibility, and repeated use in unrelated tests. I do not think going forward we should remove Mockito for this app, sense it does provide others with the option to choose. Looking for any and all feedback!

File: test/domain/repository/auth/auth_repository_fake.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import 'package:flap_app/domain/entity/request_response.dart';
+import 'package:flap_app/domain/entity/user_info.dart';
+import 'package:flap_app/domain/repository/auth/auth_repository.dart';
+
+/// A fake implementation of the [AuthRepository]. This allows us to test the [LogInUseCase] without,
+/// having to worry about the network (if one were to be used). Feel free to make any changes to
+/// fit your needs.
+class AuthRepositoryFake extends AuthRepository {
+  Future<RequestResponse<UserInfo, LoginError>> response =
+      Future.value(const Success(UserInfo(name: "", email: "")));
```
Comment (cxw320) [inline test/domain/repository/auth/auth_repository_fake.dart:12]: Maybe rename to successResponse for clarity?

File: test/domain/usecase/log_in_usecase_fake.dart [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      Future.value(const Success(UserInfo(name: "", email: "")));
+
+  void changeResponse(Future<RequestResponse<UserInfo, LoginError>> response) {
+    this.response = response;
+  }
+
+  @override
+  Future<RequestResponse<UserInfo, LoginError>> logIn(
+      LoginInformation loginInformation) {
+    return response;
```
Comment (cxw320) [inline test/domain/usecase/log_in_usecase_fake.dart:20]: same here - returning successResponse might be clearer

File: test/domain/repository/auth/auth_repository_fake.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  void changeResponse(Future<RequestResponse<UserInfo, LoginError>> response) {
+    this.response = response;
+  }
```
Comment (cxw320) [inline test/domain/repository/auth/auth_repository_fake.dart:16]: this "customization" method is really nice 👀

File: test/presentation/screens/login/notifier/login_screen_state_notifier_test.dart [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+import '../../../../domain/usecase/log_in_usecase_fake.dart';
+import '../../../../util/listener.dart';
+import '../../../../util/provider_container.dart';
+
+void main() {
+  final loginUseCase = LoginUseCaseFake();
+
+  group("Unit Tests surrounding LoginStateNotifier", () {
+    test(
+      "given loginScreenStateNotifier, with no actions taken, then initial state should be returned",
```
Comment (cxw320) [inline test/presentation/screens/login/notifier/login_screen_state_notifier_test.dart:18]: Is there a reason why the test naming convention is different here for the notifier vs the LoginUseCase? 
Screenshot 2023-10-24 at 11 26 16 AM 
I can see this also being organized as "Test the usage of LoginStateNotifier" ..."when no actions are taken, initial state should be returned" ..."when login is called, LoginError.emptyEmail returned, then correct states should be returned". Let me know which convention you prefer and I'll change my unit tests to be consistent!

File: README.md (hunk 1 above)
Comment (cxw320) [reply README.md:29]: Could we make a section on **Unit Testing Providers with Riverpod** and cite this? https://riverpod.dev/docs/essentials/testing <- this was helpful in understanding where the provider_container util class came from, and your use of stateListener in unit tests - these are all new concepts that I'm trying to wrap my head around still.

File: test/util/provider_container.dart [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+/// A testing utility which creates a [ProviderContainer] and automatically
+/// disposes it at the end of the test.
+ProviderContainer createContainer({
+  ProviderContainer? parent,
+  List<Override> overrides = const [],
+  List<ProviderObserver>? observers,
+}) {
```
Comment (cxw320) [inline test/util/provider_container.dart:11]: Maybe add in the comment above this link if others want to learn more: https://riverpod.dev/docs/essentials/testing

**Review by cxw320 — COMMENTED** [review]
In the future, I'll group all my comments in reviews (I've been doing some one-off comments by accident). This PR was incredibly helpful for learning the new riverpod 2.0 and also for seeing the power of fakes - (specifically the power of riverpod's dependency injection to override providers). I'm going to spend a bit more time with the stateListener in your loginScreenNotifier test - that's still confusing to me and I'll comment more questions if I run into any.

**Review by iamricardo — APPROVED** [review]
Looks good to me. As a note for us probably need to make sure we stick to the RiverPod v2 way from now on.

File: test/data/source/network/spoonacular_api_test.dart [hunk 8]
```diff
@@ -22,7 +22,7 @@ void main() {
         'returns recipe list and a success request response object upon 200 code',
         () async {
       dioAdapter.onGet(
-        "",
+        "https://api.spoonacular.com/recipes/random?number=20&apiKey=0793e5849b644a0d8cb4876de125a6d5",
```
Comment (cxw320) [inline test/data/source/network/spoonacular_api_test.dart:25]: Could we replace this with: ` spoonacularApi.spoonacularUri.toString(),` so api key is not hard-coded in?

Comment (cxw320) [reply test/data/source/network/spoonacular_api_test.dart:25]: I just realized that this isn't a class member yet, in my PR I did this so I could get access to the urI - feel free to take from this (not the Flavor.getBaseUrlHost - but everything else)
Screenshot 2023-10-27 at 10 37 22 AM

Comment (bcarterterp) [reply test/data/source/network/spoonacular_api_test.dart:25]: I think something got wonky by merging in `main`. I have fixed the test to use `spoonacularApi.spoonacularUri.toString()` now. I will let you're PR do the rest as to not have potential conflicts!

**Review by cxw320 — APPROVED** [review]
Looks ready to me!!

**Pull request merged**